### PR TITLE
fix: improve performance by not resetting the environment

### PR DESCRIPTION
**Pull request by SophieDePaula**
[Trello card](https://trello.com/c/rtnHxK9W/1788-yahoo-test-batch-duration-increase-after-sdk-upgrade)

This PR address a performance hit that was raised in the card above.
From my investigation I see that it is related to retrieving the `userAgent` using a snippet (it takes about a second to run , poll and get the result). We call this snippets twice on every `open` and twice on every `check`, in `core` and `core/ufg`.

I added an improvement that  should not reset the environment in `ufg` `open` and `check` as there is no operation that could be done by the user to change the context of the driver.

This improved the run time by ~35% according to the tests I ran. However, the difference in run time vs the version the customer mentioned in the card (`eyes-selenium` `4.59.4`) is still very significant .

Here are the results where Y axis is seconds and X axis is the number of tests (4.66.11 - dev is the current branch in red):

image

### CODE REVIEW

**Review by kyrylodolynskyi — COMMENTED** [review]
This is indeed a great finding. I see you are trying to solve just your problem, please go all the way in solving this problem in general, I believe this could be seen not only in ufg but classic as well, maybe in some other places.
And about the flag, I think it should work the opposite way. so we can do a full refresh only when needed (e.g. on check)

File: js/packages/driver/src/driver.ts [hunk 1]
```diff
@@ -561,7 +563,7 @@ export class Driver<T extends SpecType> {
     }
     try {
       await this._spec.switchWorld(this.target, name)
-      this.refresh()
+      this.refresh(false)
```
[the remark this answers is not in the document]
Comment (SophieDePaula) [reply js/packages/driver/src/driver.ts]: It was `false` when the flag was `noRefresh`, but now it's `isRefresh` so it's `true` as it's the default value and we don't refresh only from specific places where we know it's safe not rerefresh.

**Comment by SophieDePaula** [comment]
> This is indeed a great finding. I see you are trying to solve just your problem, please go all the way in solving this problem in general, I believe this could be seen not only in ufg but classic as well, maybe in some other places. And about the flag, I think it should work the opposite way. so we can do a full refresh only when needed (e.g. on check)

I flipped the flag and now I send `false` from `ufg` and `classic` in `open` `check` and `check-and-close`.
Do you think we should also not refresh on `core` open?

**Comment by SophieDePaula** [comment]
After the additional changes and the merge from master , I now see significant improvements in performance (not sure why).
@kyrylodolynskyi I think we can merge this, what do you think?

**Comment by kyrylodolynskyi** [comment]
@SophieDePaula I spotted a bug there, you didn't handle the fact that the on the core level between `openEyes` and `check` there is driver sharing happening, and this is an important part to reset the driver state. In general, I came up to the conclusion that reset of the state should happen only when a driver from `openEyes` is used in `check`/`checkAndClose`.
Also `refresh` method is `isRefresh` options just didn't align with me, because it does not have any explanatory properties, I would assume that `.refresh({isRefresh: false})` doing nothing. So I decided to rename it to `reset`.
I would like you to run your benchmarks with the changes I've done, I know it works right in terms of logic, but I just want to be sure that we still getting the great performance boost you discovered.

File: js/packages/driver/src/driver.ts [hunk 2]
```diff
@@ -226,7 +226,8 @@ export class Driver<T extends SpecType> {
   async getUserAgent({force}: {force?: boolean} = {}): Promise<UserAgent | undefined> {
     if (this._driverInfo?.userAgent === undefined || force) {
       this._driverInfo ??= {}
-      this._driverInfo.userAgent ??= (await this.currentContext.executePoll(snippets.getUserAgent)) ?? null
+      this._driverInfo.userAgent ??=
+        (await this.currentContext.executePoll(snippets.getUserAgent, {pollTimeout: 100})) ?? null
```
Comment (SophieDePaula) [inline js/packages/driver/src/driver.ts:230]: @kyrylodolynskyi fyi, I added this change and now the run time is the same as the old version the customer mentioned in his card (5.59.4).
There is one test in `core` that is failing, I am not sure why. It's also failing on `master` so I don't think it's related to my changes.

**Pull request merged**